Sources/EmbraceCore/SwiftUI/EmbraceTracePhase.swift

```diff
@@ -0,0 +1,234 @@
+            return nil
+        }
+        
+        // TODO: Figure out how we name things
+        let sanitizedName = "\(name.lowercased())"
+        
+        let builder = client.buildSpan(
+            name: sanitizedName,
+            type: .performance,
+            attributes: attributes ?? [:],
+            autoTerminationCode: nil
+        )
+        
+        // get the right storage for this span
+        let storage: LinearCollection = if isCycle { cycleSpans } else { spans }
+        
+        // This simply sets the new span parent to
+        // 1. the storage type current span if there is one.
+        // 2. The current cycled storage span if this span isn't cycled.
+        if let parent = storage.peek() {
+            
+            // if this storage has spans already,
+            // just make the current one this spans parent.
+            builder.setParent(parent)
+            
+        } else if !isCycle, let parent = cycleSpans.peek() {
+            
+            // if this isn't a cycled span and
+            // cycled spans are open, then make that the parent.
+            builder.setParent(parent)
+        }
+        
+        let span = builder.startSpan()
+        storage.push(span)
+        
+        return span
+    }
+    
+    /// Underlying implementation to end a span and validate stack consistency.
+    ///
+    /// - Parameters:
+    ///   - span: The span to end.
+    ///   - isCycle: Indicates if the span was a cycle span.
+    ///   - function: Calling function for diagnostics.
+    func endSpan(_ span: OpenTelemetryApi.Span?, isCycle: Bool, _ function: StaticString = #function) {
+        
+        dispatchPrecondition(condition: .onQueue(.main))
+        guard let span else {
+            logger?.error("No span passed to `endSpan()`, there might be an error in your code if a nil span is received.")
+            return
+        }
+        
+        let storage: LinearCollection = if isCycle { cycleSpans } else { spans }
+        guard let found = storage.peek() else {
+            logger?.error("Span cache is empty, are you sure you created this span using `startSpan()`?")
+            return
+        }
+        
+        guard found.context.spanId == span.context.spanId else {
+            logger?.error("No span equivalent to this span found in cache, did you create this span using `startSpan()`?")
+            return
+        }
+        
+        storage.pop()?.end()
+    }
+}
+
+/// Protocol for simple span storage collections (stack or queue).
+fileprivate protocol LinearCollection {
+    func push(_ value: Span)
+    func pop() -> Span?
+    func peek() -> Span?
+    var isEmpty: Bool { get }
+}
+
+/// Default implementation for `isEmpty` based on `peek()`.
+fileprivate extension LinearCollection {
+    var isEmpty: Bool { peek() == nil }
+}
+
+/// FIFO queue implementation for cycle-based spans.
+fileprivate class Queue: LinearCollection {
+    var storage: [Span] = []
+    func pop() -> Span? {
```

Comment:
does this function need synchronization given that we're both reading from and writing to `storage`?

Reply:
It doesn’t. Consider it to be just like a low level data structure, the sync will happen at a higher level depending on usage.